x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -65,76 +63,67 @@
-        User user,
-        RealmRef authenticatedBy,
-        RealmRef lookedUpBy,
-        Version version,
-        AuthenticationType type,
-        Map<String, Object> metadata
-    ) {
-        this.user = Objects.requireNonNull(user);
-        this.authenticatedBy = Objects.requireNonNull(authenticatedBy);
-        this.lookedUpBy = lookedUpBy;
-        this.version = version;
+    private Authentication(Subject subject, AuthenticationType type) {
+        this(subject, subject, type);
+    }
+
+    private Authentication(Subject effectiveSubject, Subject authenticatingSubject, AuthenticationType type) {
+        this.effectiveSubject = effectiveSubject;
+        this.authenticatingSubject = authenticatingSubject;
         this.type = type;
-        this.metadata = metadata;
-        if (user instanceof RunAsUser runAsUser) {
-            authenticatingSubject = new Subject(runAsUser.authenticatingUser, authenticatedBy, version, metadata);
-            // The lookup user for run-as currently doesn't have authentication metadata associated with them because
-            // lookupUser only returns the User object. The lookup user for authorization delegation does have
-            // authentication metadata, but the realm does not expose this difference between authenticatingUser and
-            // delegateUser so effectively this is handled together with the authenticatingSubject not effectiveSubject.
-            effectiveSubject = new Subject(user, lookedUpBy, version, Map.of());
-        } else {
-            authenticatingSubject = effectiveSubject = new Subject(user, authenticatedBy, version, metadata);
-        }
-        this.assertApiKeyMetadata();
-        this.assertDomainAssignment();
+        assertInternalConsistency();
     }
 
     public Authentication(StreamInput in) throws IOException {
-        this.user = AuthenticationSerializationHelper.readUserFrom(in);
-        this.authenticatedBy = new RealmRef(in);
+        // Read the user(s)
+        final User outerUser = AuthenticationSerializationHelper.doReadUserFrom(in);
+        final boolean hasInnerUser;
+        if (User.isInternal(outerUser)) {
+            hasInnerUser = false;
+        } else {
+            hasInnerUser = in.readBoolean();
+        }
+        final User innerUser;
+        if (hasInnerUser) {
+            innerUser = AuthenticationSerializationHelper.readUserFrom(in);
+            assert false == User.isInternal(innerUser) : "internal users cannot participate in run-as";
+        } else {
+            innerUser = null;
+        }
+
+        final RealmRef authenticatedBy = new RealmRef(in);
+        final RealmRef lookedUpBy;
         if (in.readBoolean()) {
-            this.lookedUpBy = new RealmRef(in);
+            assert innerUser != null : "inconsistent looked-up-by realm and inner-user";
+            lookedUpBy = new RealmRef(in);
         } else {
-            this.lookedUpBy = null;
+            lookedUpBy = null;
```

Comment:
> Do we need an `assert innerUser == null` here?

Unfortunately no. We use `null` lookedUpBy to indicate run-as failure (when the run-as user does not exists). In this case, we will have non-null innerUser but `null` lookedUpBy.

But the reverse is true, i.e. if innerUser is `null`, we should definitely have `null` lookedUpBy as well. That's why the other assertion is down below because it does not fit here. 

That said, I pulled them together into a single statement and added an elaborative comment for it.

Reply:
Tweaked it a bit and made it simpler.